activerecord/test/cases/autosave_association_test.rb

```diff
@@ -302,6 +302,26 @@ def test_callbacks_firing_order_on_save
     eye.update(iris_attributes: { color: "blue" })
     assert_equal [false, false, false, false], eye.after_save_callbacks_stack
   end
+
+  def test_callbacks_on_child_when_parent_autosaves_child
+    eye = Eye.create(iris: Iris.new)
+    assert_equal 1, eye.iris.before_validation_callbacks_counter
+    assert_equal 1, eye.iris.before_create_callbacks_counter
+    assert_equal 1, eye.iris.before_save_callbacks_counter
+    assert_equal 1, eye.iris.after_validation_callbacks_counter
+    assert_equal 1, eye.iris.after_create_callbacks_counter
+    assert_equal 1, eye.iris.after_save_callbacks_counter
+  end
+
+  def test_callbacks_on_child_when_child_autosaves_parent
+    iris = Iris.create(eye: Eye.new)
+    assert_equal 1, iris.before_validation_callbacks_counter
+    assert_equal 1, iris.before_create_callbacks_counter
+    assert_equal 1, iris.before_save_callbacks_counter
+    assert_equal 1, iris.after_validation_callbacks_counter
+    assert_equal 1, iris.after_create_callbacks_counter
+    assert_equal 1, iris.after_save_callbacks_counter
```

Comment:
Before patch:
<img width="1133" alt="Screenshot 2023-10-30 at 6 21 11 pm" src="https://github.com/rails/rails/assets/54629302/a4e628ac-eba1-4fd2-b67a-50c40e6d8cf3">

**Note:** The `before_create` and `after_create` cases worked fine before the patch, I only added it in for consistency and to catch regressions.